File: exercise/ts/day03/src/gift.ts
export class Gift {
    private attributes: Map<string, string> = new Map<string, string>();

    constructor(
        private readonly name: string,
        private readonly weight: number,
        private readonly color: string,
        private readonly material: string,
    ) {
    }

    public assignToChild(childName: string): void {
    }

    public addAttribute(key: string, value: string): void {
        this.attributes.set(key, value);
    }

    public getRecommendedAge(): number {
        const recommendedAge = this.attributes.get('recommendedAge');
        return recommendedAge ? parseInt(recommendedAge, 10) : 0;
    }

    public toString(): string {
        return `A ${this.color}-colored ${this.name} weighing ${this.weight} kg made in ${this.material}`;
    }
}

File: solution/day03/src/main/java/preparation/Gift.java
package preparation;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import static java.lang.String.format;

public class Gift {
    private final String name;
    private final double weight;
    private final String color;
    private final String material;
    private final Map<String, String> attributes = new HashMap<>();
    private String childName;

    public Gift(String name, double weight, String color, String material) {
        this.name = name;
        this.weight = weight;
        this.color = color;
        this.material = material;
    }

    public void assignToChild(String childName) {
        this.childName = childName;
    }

    public void addAttribute(String key, String value) {
        attributes.put(key, value);
    }

    private Optional<String> getAttribute(String key) {
        return Optional.ofNullable(attributes.get(key));
    }

    public int getRecommendedAge() {
        return getAttribute("recommendedAge")
                .map(Integer::parseInt)
                .orElse(0);
    }

    @Override
    public String toString() {
        return format("A %s-colored %s weighting %s kg made in %s", color, name, weight, material);
    }
}